File: wasm/sample/src/xplpc/proxy/platform-proxy.ts
import { XPLPC } from "../core/xplpc";
import { MappingList } from "../data/mapping-list";
import { ExceptionMessage } from "../util/exception-message";
import { Log } from "../util/log";

interface PlatformProxy {
    initialize(): void;
    initializePlatform(): void;
    callProxy(key: string, data: string): void;
    hasMapping(name: string): boolean;
}

function callNativeProxyCallback(key: string, data: string): void {
    XPLPC.shared().module.CallbackList.executeFromJavascript(key, data);
}

const platformProxy: PlatformProxy = {
    initialize() {
        this.initializePlatform();
    },

    initializePlatform() {},

    callProxy(key: string, data: string) {
        if (!XPLPC.shared().initialized) {
            Log.e("[PlatformProxy : call] XPLPC was not initialized");
            callNativeProxyCallback(key, "");
            return;
        }

        const request = XPLPC.shared().config.serializer.decodeRequest(data);

        if (!request) {
            callNativeProxyCallback(key, "");
            return;
        }

        if (!request.functionName) {
            Log.e("[PlatformProxy : call] Function name is empty");
            callNativeProxyCallback(key, "");
            return;
        }

        const mappingItem = MappingList.shared().find(request.functionName);

        if (!mappingItem) {
            Log.e(
                "[PlatformProxy : call] Mapping not found for function: " +
                    request.functionName,
            );
            callNativeProxyCallback(key, "");
            return;
        }

        mappingItem
            .target(request.message)
            .then((response: unknown) => {
                callNativeProxyCallback(
                    key,
                    XPLPC.shared().config.serializer.encodeFunctionReturnValue(
                        response,
                    ),
                );
            })
            .catch((e: unknown) => {
                Log.e(
                    '[PlatformProxy : call] Error when execute function "' +
                        request.functionName +
                        '"',
                );
                Log.d(
                    '[PlatformProxy : call] Error when execute function "' +
                        request.functionName +
                        '": ' +
                        ExceptionMessage.get(e),
                );
                callNativeProxyCallback(key, "");
            });
    },

    hasMapping(name: string): boolean {
        return MappingList.shared().has(name);
    },
};

export default platformProxy;

File: kotlin/shared/src/com/xplpc/proxy/PlatformProxy.kt
package com.xplpc.proxy

import com.xplpc.core.XPLPC
import com.xplpc.data.CallbackList
import com.xplpc.data.MappingList
import com.xplpc.message.Message
import com.xplpc.util.Log

class PlatformProxy private constructor() {
    companion object {
        @JvmStatic
        external fun callNativeProxy(key: String, data: String)

        @JvmStatic
        external fun callNativeProxyCallback(key: String, data: String)

        @JvmStatic
        fun onNativeProxyCallback(key: String, data: String) {
            CallbackList.execute(key, data)
        }

        @JvmStatic
        fun onNativeProxyCall(key: String, data: String) {
            if (!XPLPC.initialized) {
                Log.e("[PlatformProxy : call] XPLPC was not initialized")
                callNativeProxyCallback(key, "")
                return
            }

            val request = XPLPC.config.serializer.decodeRequest(data)

            if (request == null) {
                callNativeProxyCallback(key, "")
                return
            }

            if (request.functionName.isEmpty()) {
                Log.e("[PlatformProxy : call] Function name is empty")
                callNativeProxyCallback(key, "")
                return
            }

            val mappingItem = MappingList.find(request.functionName)

            if (mappingItem == null) {
                Log.e("[PlatformProxy : call] Mapping not found for function: ${request.functionName}")
                callNativeProxyCallback(key, "")
                return
            }

            execute(key, request.functionName, mappingItem.target, request.message)
        }

        @JvmStatic
        fun onHasMapping(name: String): Boolean {
            return MappingList.has(name)
        }

        @JvmStatic
        fun onInitializePlatform() {
        }

        @JvmStatic
        fun onFinalizePlatform() {
            MappingList.clear()
        }

        private fun execute(
            key: String,
            functionName: String,
            target: (Message, (Any?) -> Unit) -> Unit,
            message: Message,
        ) {
            try {
                target(message) { response ->
                    val encodedData = try {
                        XPLPC.config.serializer.encodeFunctionReturnValue(response)
                    } catch (e: Exception) {
                        Log.e("[PlatformProxy : call] Error when encode data")
                        Log.d("[PlatformProxy : call] Error when encode data: ${e.message}")
                        ""
                    }

                    callNativeProxyCallback(key, encodedData)
                }
            } catch (e: Throwable) {
                // An error is not an exception here, and letting one leave this frame makes it a pending exception at the jni boundary with the caller never answered.
                Log.e("[PlatformProxy : call] Error when execute function \"$functionName\"")
                Log.d("[PlatformProxy : call] Error when execute function \"$functionName\": ${e.message}")
                callNativeProxyCallback(key, "")
            }
        }
    }
}
